wave_utils/include/wave/utils/pose_cov_comp.hpp

```diff
@@ -0,0 +1,233 @@
+ *
+ *  @return composed pose with predicted covariance
+ *
+ */
+PoseWithCovariance composePose(PoseWithCovariance &p1, PoseWithCovariance &p);
+
+/** The Jacobian of quaternion normalization function. Quaternion in the form of
+ *  [qr, qx, qt, qz]
+ *  Equation (1.7)
+ *
+ *  @param q quaternion vector
+ *
+ *  @return quaternion normal wrt q Jacobian
+ */
+Matrix4x4 jacobian_Quat_Norm_wrt_q(const Vector4 &q);
+
+/** The Jacobian of normalized quaternion to rpy function.
+ *  Equation (2.9) to Equation (2.10)
+ *
+ *  @param q quaternion vector
+ *
+ *  @return quaternion normal to roll pitch yaw wrt q Jacobian
+ */
+Matrix3x4 jacobian_Quat_Norm_to_Rpy_wrt_q(const Vector4 &q);
+
+/** The Jacobian of p7 to p6 conversion.
+ *  Equation (2.12)
+ *
+ *  @param p pose with position and quaternion
+ *
+ *  @return p7 to p6 wrt p Jacobian
+ */
+Matrix6x7 jacobian_p7_to_p6_wrt_p(const Vector7 &p);
+
+/** Jacobian of composing a point to a p7.
+ *  Equation (3.8)
+ *
+ *  @param p pose with position and quaternion
+ *  @param a position
+ *  @return p7 point composition wrt p Jacobian
+ */
+Matrix3x7 jacobian_p7_Point_Composition_wrt_p(const Vector7 &p,
+                                              const Vector3 &a);
+/** Jacobian of the composition of p7 poses.
+ *  Equation (5.8)
+ *
+ *  @param p1 pose with position and quaternion
+ *  @param p2 pose with position and quaternion
+ *  @return p7 p7 composition wrt p1 Jacobian
+ */
+Matrix7x7 jacobian_p7_p7_Composition_wrt_p1(const Vector7 &p1,
+                                            const Vector7 &p2);
+
+/** Jacobian of composing a point to a p7.
+ *  Equation (3.10)
+ *
+ *  @param p pose with position and quaternion
+ *  @param a position
+ *  @return p7 point composition wrt a Jacobian
+ */
+Matrix3x3 jacobian_p7_Point_Composition_wrt_a(const Vector7 &p,
+                                              const Vector3 &a);
+
+/** Jacobian of the composition of p7 poses.
+ *  Equation (5.9)
+ *
+ *  @param p1 pose with position and quaternion
+ *  @param p2 pose with position and quaternion
+ *
+ *  @return p7 p7 composition wrt p2 Jacobian
+ */
+Matrix7x7 jacobian_p7_p7_Composition_wrt_p2(const Vector7 &p1,
+                                            const Vector7 &p2);
+
+/** Jacobian of converting a p6 to a p7.
+ *  Equation (2.8)
+ *
+ *  @param p pose with position and ypr
+ *  @return p6 to p7 wrt p Jacobian
+ */
+Matrix7x6 jacobian_p6_to_p7_wrt_p(const Vector6 &p);
+}
+
+/** NOTE: wave_geometry/ has a similar implementation of the following functions
+ *  but these are made to match the methods proposed in [1]. These functions
+ *  should ONLY be used for pose composition based on [1].
+ *
+ */
+namespace wave {
+namespace pose_comp {
+
+/** Normalizes the given Quaternion.
+ *  Uses the equation qr^2 + qx^2 + qy^2 + qz^2 = 1.
+ *  Equation(1.6)
+ *
+ *  @param q quaternion vector
+ *
+ *  @return normalized quaternion
+ */
+Vector4 normalizeQuat(const Vector4 q);
```

Comment:
missing &

Reply:
I realize you are just copying this from another repo, but my comments from there still apply. There is an `Eigen::Quaternion` class which already provides a `normalize` function. We should not provide the same functionality with a worse interface. Use `Quaternion` instead of `Vector4d`.

Reply:
Converted quaternion vector representations to `Eigen::Quaterniond`